### PR TITLE
Layout component - coldiv.

### PR DESCRIPTION
**Pull request by mahidhar-reddy09**
Set up layout component- coldiv.

### CODE REVIEW

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/util/utility-types.ts [hunk 1]
```diff
@@ -10,3 +12,34 @@ export type RequiredKeys<T> = {
...
+    ? [...Result, Result["length"]]["length"]
+    : AddOne<N, [...Result, Result["length"]]>;
+
+// uses the range and if there is no max defined it just becomes a number
+export type ColSpan<Max extends number | undefined> = Max extends number
+    ? Range<Max> | [Range<AddOne<Max>>, Range<AddOne<Max>>] | undefined
+    : number | [number, number] | undefined;
+
+// using generic breakpointspan to extract column span for a specific breakpoint
+export type BreakpointSpan<
```
Comment (qroll) [inline src/util/utility-types.ts:38]: sorry! ColSpan and Breakpoint can remain in the layout types as they're related to the component

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/layout/types.ts [hunk 2]
```diff
@@ -14,3 +17,49 @@ export interface ContainerProps extends CommonLayoutProps {
...
+    xsCols?: BreakpointSpan<"xs">;
+    xxsCols?: BreakpointSpan<"xxs">;
+}
+export interface ColDivProps
+    extends React.HTMLAttributes<HTMLDivElement>,
+        ColProps {
+    "data-testid"?: string;
+
+    xxlMargin?: string | number;
+    xxlGutter?: string | number;
```
Comment (qroll) [inline src/layout/types.ts:52]: to remove these props as well

File: src/layout/col-div.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            $smStart: smStartSpan.start,
+            $smSpan: smStartSpan.span,
+            $xsStart: xsStartSpan.start,
+            $xsSpan: xsStartSpan.span,
+            $xxsStart: xxsStartSpan.start,
+            $xxsSpan: xxsStartSpan.span,
+
+            $xxlMargin: BreakpointValues["xxl-margin"]({ theme }),
+            $xxlGutter: BreakpointValues["xxl-gutter"]({ theme }),
+            $xlMargin: BreakpointValues["xl-margin"]({ theme }),
```
Comment (qroll) [inline src/layout/col-div.tsx:92]: these can be removed if they are unused

File: src/util/utility-types.ts (hunk 1 above)
Comment (mahidhar-reddy09) [reply src/util/utility-types.ts:38]: Very interesting actually i moved them to the utility and realised actually they are related to the component and so I moved them back to types. But seems when I commited it was captured as tho it was still in utils...

Comment (qroll) [reply src/util/utility-types.ts:38]: do revert the changes in this file accordingly

File: src/layout/col-div.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        const xsColumnCount = BreakpointValues["xs-column"]({ theme });
+        const xxsColumnCount = BreakpointValues["xxs-column"]({ theme });
+
+        const xxlStartSpan = getColSpan(xxlCols, xxlColumnCount, "xxl");
+        const xlStartSpan = getColSpan(xlCols, xlColumnCount, "xl");
+        const lgStartSpan = getColSpan(lgCols, lgColumnCount, "lg");
+        const mdStartSpan = getColSpan(mdCols, mdColumnCount, "md");
+        const smStartSpan = getColSpan(smCols, smColumnCount, "sm");
+        const xsStartSpan = getColSpan(xsCols, xsColumnCount, "xs");
+        const xxsStartSpan = getColSpan(xxsCols, xxsColumnCount, "xxs");
```
Comment (qroll) [inline src/layout/col-div.tsx:72]: the original implementation also supports fallback behaviour e.g. desktop will automatically inherit tablet if not specified

to restore this functionality

File: src/layout/col-div.tsx [hunk 5]
```diff
@@ -63,12 +62,63 @@ const Component = (
...
+                smCols ||
+                xsCols ||
+                xxsCols,
+            xxlColumnCount ||
+                xlColumnCount ||
+                lgColumnCount ||
+                mdColumnCount ||
+                smColumnCount ||
+                xsColumnCount ||
+                xxsColumnCount,
```
Comment (qroll) [inline src/layout/col-div.tsx:79]: column count doesn't need a fallback as it's static (based on the theme)

**Review by qroll — CHANGES_REQUESTED** [review]

**Review by qroll — APPROVED** [review]

**Pull request merged**